onnxruntime/contrib_ops/cpu/bert/multihead_attention.cc

```diff
@@ -138,6 +144,40 @@ Status MultiHeadAttention<T>::Compute(OpKernelContext* context) const {
   ORT_RETURN_IF_ERROR(MaybeTransposeToBNSHAndAddBias<T>(
       context, allocator, batch_size, num_heads_, kv_sequence_length, v_head_size, value, bias, v_bias_offset, V));
 
+  if (std::is_same_v<T, float> && !disable_flash_ && !is_unidirectional_ && key_padding_mask == nullptr && extra_add_qk == nullptr && past_key == nullptr && past_value == nullptr && present_k == nullptr && present_v == nullptr) {
+    FlashAttentionThreadedArgs args;
+    args.batch_size = batch_size;
+    args.num_heads = num_heads_;
+    args.q_sequence_length = q_sequence_length;
+    args.kv_sequence_length = kv_sequence_length;
+    args.qk_head_size = qk_head_size;
+    args.v_head_size = v_head_size;
+    args.scale = (scale_ == 0.0f) ? 1.0f / sqrt(static_cast<float>(qk_head_size)) : scale_;
+
+    const auto& env = Env::Default();
+    int l2_cache_size = env.GetL2CacheSize();
+    args.row_size_kv = l2_cache_size / sizeof(float) / 4 / (qk_head_size + v_head_size);
```

Comment:
Add some comments why the formula is row_size_kv = l2_cache_size / sizeof(float) / 4 / (qk_head_size + v_head_size). For example, there is 4 and sizeof(float) at same time.

When there are multiple threads, all threads will use L2 cache. For example, CPU has 4MB L2 cache, head size is 64, then row_size_kv is 2048. That's a very large number.

Reply:
I added some comments. Basically, we pick these values so that the sliced matrices can fit in L2 cache.
I added limits: `row_size_kv <= kv_seq_len` and `row_size_q <= q_seq_len`. They are used to slice the inputs along seq_len so there is no point to have them larger than seq_len. This can also prevent allocating excessive memory when seq_len is small.

Reply:
The `l2_cache_size_` stores L2 cache size for a single core. In recent x86-64 architectures, the L2 cache is usually not shared among physical cores. I don't know if this is true for other architectures.
ONNXRuntime usually allocates one thread per physical core so we don't need to divide it by number of threads.
However, if the number of threads is larger than the number of physical cores, we might see a performance degradation because different threads will start competing for the L2 space.